src/mesh/domain.hpp

```diff
@@ -0,0 +1,172 @@
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef MESH_DOMAIN_HPP_
+#define MESH_DOMAIN_HPP_
+
+#include <array>
+
+#include "athena.hpp" 
+#include "athena_arrays.hpp"
+
+namespace parthenon {
+
+  /// Defines the maximum size of the static array used in the IndexShape objects
+  const int NDIM = 3;
+
+  struct IndexRange {
+    int start = 0; /// Starting Index (inclusive)
+    int end = 0; /// Ending Index (inclusive)
+    int ncells() const noexcept { return end-start+1; }
+  };
+  
+  // Assuming we have a block
+  //
+  //  - - - - - - - - - -   ^
+  //  |  |  ghost    |  |   | 
+  //  - - - - - - - - - -   | 
+  //  |  |     ^     |  |   | 
+  //  |  |     |     |  |   
+  //  |  | interior  |  |   entire
+  //  |  |     |     |  |    
+  //  |  |     v     |  |   | 
+  //  - - - - - - - - - -   |
+  //  |  |           |  |   | 
+  //  - - - - - - - - - -   v 
+  //
+  enum IndexShapeType {
+    entire,
+    interior
+  };
+
+  //! \class IndexVolume
+  //  \brief Defines the dimensions of a shape of indices
+  //
+  //  Defines the range of each dimension of the indices by defining a starting and stopping index
+  //  also contains a label for defining which region the index shape is assigned too 
+  class IndexShape {
+    private:
+      std::array<IndexRange,NDIM> x_;
+      std::array<int,NDIM> entire_ncells_;
+
+    public:
+  
+      IndexShape() {};
+
+      IndexShape(const int & nx1, const int & nx2, const int & nx3, const int & ng) 
+      : IndexShape( std::vector<int> {nx1,nx2,nx3}, ng) {};
+
+      IndexShape(const int & nx1, const int & nx2, const int & ng) 
+      : IndexShape( std::vector<int> {nx1,nx2}, ng) {};
+
+      IndexShape(const int & nx1, const int & ng) 
+      : IndexShape( std::vector<int> {nx1}, ng) {};
+
+      IndexShape(const std::vector<int> & interior_dims, const int & ng) { 
+        assert(interior_dims.size()<=NDIM && "IndexShape cannot be initialized, the number of "
+              "dimensions exceeds the statically set dimensions, you will need to change the NDIM "
+              "constant.");
+        for( int dim=1, index=0; dim<=NDIM; ++dim, ++index){
+          if (dim <= interior_dims.size()) {
+            assert( interior_dims.at(index) > 0 && "IndexShape cannot be initialized with fewer "
+                "than 1 interior cells for each dimension");
+            x_[index].start = ng;
+            x_[index].end = x_[index].start + interior_dims.at(index) - 1;
+            entire_ncells_[index] = interior_dims.at(index) + 2*ng;
+          } else {
+            x_[index].start = 0;
+            x_[index].end = 0;
+            entire_ncells_[index] = 1;
+          }
+        }
+      };
+
+      inline int x1s(const IndexShapeType & type) const 
```

Comment:
These names are an improvement. One small thing that concerns me is that `x1` represents the coordinates. I'm not sure if it's confusing to  use `x1s` and `x1e` here, since that's kind of oversubscribing the name. This could just be `is` and `ie`, which would be unambiguous in this context.